algorithms/src/Kokkos_Random.hpp

```diff
@@ -901,52 +901,51 @@ class Random_XorShift64_Pool {
   KOKKOS_DEFAULTED_FUNCTION Random_XorShift64_Pool& operator=(
       Random_XorShift64_Pool const&) = default;
 #else
-  Random_XorShift64_Pool()   = default;
+  Random_XorShift64_Pool() = default;
 #endif
+
   Random_XorShift64_Pool(uint64_t seed) {
     num_states_ = 0;
 
     init(seed, execution_space().concurrency());
   }
```

Comment:
I would delegate to the constructor that takes an exec space argument
```suggestion
  Random_XorShift64_Pool(uint64_t seed) : Random_XorShift64_Pool(execution_space(), seed) {
  }
```
(not formatted)

Reply:
Sounds good to me. That's what I had done initially, but then I had some doubts that I had perhaps made unwarranted assumptions.

Reply:
With the calls to `deep_copy()` in `init()`, we want a fence for the default execution space instance, but not otherwise. On the other hand, we'd be always calling the three arg version of `deep_copy` with a delegated ctor, so I'm not sure how to differentiate the two cases (in the delegated ctor case).

Reply:
Good point.  We could have a trailing parameter (something like strong enumeration) to denote whether we need to fence at all.
```C++
namespace Impl {
  enum RandomPoolConstructorIsAsync : bool { no = false, yes = true};
}

Random_Whatev_Pool(ExecutionSpace const& exec, uint64_t seed, Impl::RandomPoolConstructorIsAsync is_async = RandomPoolConstructorIsAsync::yes) {
  // ...
  if (static_cast<bool>(is_async)) {
    exec.fence("Random_Whatev_Pool(seed) constructor sync");
  }
}
Random_Whatev_Pool(uint64_t seed) : Random_Whatev_Pool(ExecutionSpace(), seed, RandomPoolConstructorIsAsync::no) {}
```
(not tested)